finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/Stream.scala

```diff
@@ -68,6 +68,17 @@ object Stream {
   // performed).  This would require some form of buffering, ideally
   // in the dispatcher and server stream
 
+  /**
+   * Read all frames from the queue and release them.  This can be called on a queue after
+   * calling q.fail(e, discard = false) as a safe alternative to q.fail(e, discard = true).
+   */
+  def drainFrameQueue(q: AsyncQueue[Frame]): Future[Nothing] = {
+    q.poll().flatMap { f =>
```

Comment:
consider including `q.fail(e, discard = false)` in here -- since this is never called without also failing immediately prior afaict.  then can be renamed to `failFrameQueue`

Reply:
good idea